File: app/api/function_assets.py
from __future__ import annotations

from app.schemas.function_asset import FunctionAsset
from app.services.function_asset_service import function_asset_service
from fastapi import APIRouter

router = APIRouter(prefix="/api/function-assets", tags=["function-assets"])


@router.get("", response_model=list[FunctionAsset])
def list_function_assets() -> list[FunctionAsset]:
    return function_asset_service.list_assets()


@router.post("", response_model=FunctionAsset)
def create_function_asset(payload: dict) -> FunctionAsset:
    return function_asset_service.create_asset(payload)


@router.post("/import-csv", response_model=FunctionAsset)
def import_function_asset_csv(payload: dict) -> FunctionAsset:
    return function_asset_service.import_csv(payload)


@router.get("/{function_id}", response_model=FunctionAsset)
def get_function_asset(function_id: str) -> FunctionAsset:
    return function_asset_service.get_asset(function_id)


@router.put("/{function_id}", response_model=FunctionAsset)
def update_function_asset(function_id: str, payload: dict) -> FunctionAsset:
    return function_asset_service.update_asset(function_id, payload)


@router.post("/{function_id}/validate")
def validate_function_asset(function_id: str, payload: dict | None = None) -> dict:
    return function_asset_service.validate_asset(function_id, payload)


@router.post("/{function_id}/preview")
def preview_function_asset(function_id: str, payload: dict | None = None) -> dict:
    return function_asset_service.preview_asset(function_id, payload)

File: frontend/src/api/functionAssets.ts
import { apiClient, unwrap } from './client';
import type { FunctionAsset, FunctionAssetPreview, FunctionAssetValidation } from '../types/functionAsset';

export const getFunctionAssets = () => unwrap<FunctionAsset[]>(apiClient.get('/api/function-assets'));
export const getFunctionAsset = (id: string) => unwrap<FunctionAsset>(apiClient.get(`/api/function-assets/${id}`));
export const createFunctionAsset = (payload: Partial<FunctionAsset>) => unwrap<FunctionAsset>(apiClient.post('/api/function-assets', payload, { suppressErrorToast: true }));
export const importFunctionAssetCsv = (payload: Record<string, unknown>) => unwrap<FunctionAsset>(apiClient.post('/api/function-assets/import-csv', payload, { suppressErrorToast: true }));
export const updateFunctionAsset = (id: string, payload: Partial<FunctionAsset>) => unwrap<FunctionAsset>(apiClient.put(`/api/function-assets/${id}`, payload, { suppressErrorToast: true }));
export const validateFunctionAsset = (id: string, payload?: Partial<FunctionAsset>) => unwrap<FunctionAssetValidation>(apiClient.post(`/api/function-assets/${id}/validate`, payload || {}));
export const previewFunctionAsset = (id: string, payload?: Record<string, unknown>) => unwrap<FunctionAssetPreview>(apiClient.post(`/api/function-assets/${id}/preview`, payload || {}));

export const checkFunctionAssetApiReady = async () => {
  try {
    const health = await unwrap<Record<string, any>>(apiClient.get('/api/health', { suppressErrorToast: true }));
    const supports = health?.api_versions?.function_assets?.supports;
    return Array.isArray(supports) && supports.includes('POST create') && supports.includes('POST import-csv') && supports.includes('piecewise_2d');
  } catch {
    return false;
  }
};
